### PR TITLE
Replace manually crafted REST client with auto-generated one

### PR DESCRIPTION
**Pull request by Y-Sindo**
changed files can be categorized into three groups which are committed in three separate three commit:
1. `src/Common/REST/*`: files generated by Autorest
2. `src\Microsoft.Azure.SignalR.Common\REST`: enable generated REST client to authorize with JWT token.
3. other files: replace the old REST client

### CODE REVIEW

File: src/Microsoft.Azure.SignalR.Common/Generated/GeneratedRestClient.cs [hunk 1]
```diff
@@ -0,0 +1,326 @@
+// <auto-generated>
```
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Common/Generated/GeneratedRestClient.cs:1]: can auto-generated code be ignored?

File: src/Microsoft.Azure.SignalR.Common/Utilities/JwtTokenCredentials.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+// ----------------------------------------------------------------------------------
+
+using System;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Rest;
+
+namespace Microsoft.Azure.SignalR.Management
```
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Common/Utilities/JwtTokenCredentials.cs:22]: should not be `Microsoft.Azure.SignalR.Management` here

File: src/Microsoft.Azure.SignalR.Common/Utilities/RestClient.cs [hunk 3]
```diff
@@ -12,6 +12,7 @@
 using Microsoft.Azure.SignalR.Common;
 using Microsoft.Extensions.Primitives;
 using Newtonsoft.Json;
+using Microsoft.Azure.SignalR.Models;
```
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Common/Utilities/RestClient.cs:15]: Is it used?

File: src/Microsoft.Azure.SignalR.Common/Utilities/JwtTokenCredentials.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+{
+    class JwtTokenCredentials : ServiceClientCredentials
```
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Common/Utilities/JwtTokenCredentials.cs:24]: add protection level

File: src/Microsoft.Azure.SignalR.Common/Generated/GeneratedRestClient.cs (hunk 1 above)
Comment (Y-Sindo) [reply src/Microsoft.Azure.SignalR.Common/Generated/GeneratedRestClient.cs:1]: Currently the code is not generated during building. It has to be generated by Autorest which requireds Node.js environment. As the generated codes change little, I think it is more convenient to keep it here.

File: src/Microsoft.Azure.SignalR.Common/Utilities/RestClient.cs (hunk 3 above)
Comment (Y-Sindo) [reply src/Microsoft.Azure.SignalR.Common/Utilities/RestClient.cs:15]: Yes, it's used by [Payload class](https://github.com/Azure/azure-signalr/blob/00b898138dd6e0047ba174c4bf034e7c94201b30/src/Microsoft.Azure.SignalR.Common/Utilities/RestClient.cs#L128). In the past, Payload is under Microsoft.Azure.SignalR.Models. Now Payload class is generated, so the namespace is changed.

File: src/Microsoft.Azure.SignalR.Common/Utilities/JwtTokenCredentials.cs (hunk 4 above)
Comment (Y-Sindo) [reply src/Microsoft.Azure.SignalR.Common/Utilities/JwtTokenCredentials.cs]: fixed in [fix](https://github.com/Azure/azure-signalr/pull/992/commits/d58ec08f9e255560335ffc4bdf60be844ca8ee9b)

File: src/Microsoft.Azure.SignalR.Common/Utilities/JwtTokenCredentials.cs (hunk 2 above)
Comment (Y-Sindo) [reply src/Microsoft.Azure.SignalR.Common/Utilities/JwtTokenCredentials.cs]: fixed in [fix](https://github.com/Azure/azure-signalr/pull/992/commits/d58ec08f9e255560335ffc4bdf60be844ca8ee9b)

File: run.cmd [hunk 5]
```diff
@@ -1,2 +0,0 @@
-@ECHO OFF
```
Comment (vicancy) [inline run.cmd:1]: revert the change?

File: src/Microsoft.Azure.SignalR.Common/Utilities/GeneratedRestClient.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+using System;
+using System.Net.Http.Headers;
+using Microsoft.Rest;
+
+namespace Microsoft.Azure.SignalR
+{
+    public partial class GeneratedRestClient
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Utilities/GeneratedRestClient.cs:21]: a better naming, this is not Generated

Comment (Y-Sindo) [reply src/Microsoft.Azure.SignalR.Common/Utilities/GeneratedRestClient.cs:21]: This a a partial class and one part is generated.

File: src/Microsoft.Azure.SignalR.Common/REST/GeneratedRestClient.cs [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using Microsoft.Rest;
+
+namespace Microsoft.Azure.SignalR
+{
+    public partial class GeneratedRestClient
+    {
+        string ProductInfo { get; }
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.Common/REST/GeneratedRestClient.cs:12]: visibility

File: src/Common/REST/HealthApi.cs [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+            cancellationToken.ThrowIfCancellationRequested();
+            _httpResponse = await Client.HttpClient.SendAsync(_httpRequest, System.Net.Http.HttpCompletionOption.ResponseHeadersRead, cancellationToken).ConfigureAwait(false);
+            if (_shouldTrace)
+            {
+                ServiceClientTracing.ReceiveResponse(_invocationId, _httpResponse);
+            }
+            HttpStatusCode _statusCode = _httpResponse.StatusCode;
+            cancellationToken.ThrowIfCancellationRequested();
+            string _responseContent = null;
+            if ((int)_statusCode != 200)
```
Comment (wanlwanl) [inline src/Common/REST/HealthApi.cs:124]: In original logic, badgateway and serviceunavailable doesn't throw exceptions.

File: src/Microsoft.Azure.SignalR.Management/ServiceManager.cs [hunk 9]
```diff
@@ -160,27 +158,15 @@ public string GenerateClientAccessToken(string hubName, string userId = null, IL
...
-                            isHealthy = false;
-                            return true;
-                        default:
-                            return false;
-                    }
-                },
-                cancellationToken: cancellationToken);
-            return isHealthy;
+            var healthApi = _restClient.HealthApi;
+            var httpOperationResponse = await healthApi.GetHealthStatusWithHttpMessagesAsync(cancellationToken: cancellationToken);
```
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Management/ServiceManager.cs:162]: the original code has some azure exceptions like this
```
            throw response.StatusCode switch
            {
                HttpStatusCode.BadRequest => new AzureSignalRInvalidArgumentException(response.RequestMessage.RequestUri.ToString(), innerException, detail),
                HttpStatusCode.Unauthorized => new AzureSignalRUnauthorizedException(response.RequestMessage.RequestUri.ToString(), innerException),
                HttpStatusCode.NotFound => new AzureSignalRInaccessibleEndpointException(response.RequestMessage.RequestUri.ToString(), innerException),
                _ => new AzureSignalRRuntimeException(response.RequestMessage.RequestUri.ToString(), innerException),
            };
```
Is it possible to config the exceptions for autorest client? otherwise it's a breaking change.

File: src/Common/REST/HealthApi.cs (hunk 8 above)
Comment (wanlwanl) [reply src/Common/REST/HealthApi.cs:124]: add tests

File: src/Microsoft.Azure.SignalR.Common/Generated/GeneratedRestClient.cs (hunk 1 above)
Comment (wanlwanl) [reply src/Microsoft.Azure.SignalR.Common/Generated/GeneratedRestClient.cs]: So why not choose a library only requires dotnet?

What about adding a script to prepare the environment like this:
https://github.com/dotnet/aspnetcore/blob/master/docs/BuildFromSource.md

Comment (Y-Sindo) [reply src/Microsoft.Azure.SignalR.Common/Generated/GeneratedRestClient.cs]: Autorest is widely used in Azure-Powershell and azure-sdk-for-net projects and their common practice is just including the generated code into version control. I haven't found similar library like it only depending on dotnet, and if you know about this, we could have discussion. Actually I have tried to include the generation script into our build.cmd or build.sh and added Node.js environment into CI workflow but failed to start the build. Maybe I got something wrong with workflow files. 
I think another reason to upload the generated file is that we can review the generated code on Github.

File: src/Common/REST/HealthApi.cs (hunk 8 above)
Comment (Y-Sindo) [reply src/Common/REST/HealthApi.cs:124]: Yes, it's a bug. Thank you for pointing out that.

File: src/Microsoft.Azure.SignalR.Common/Generated/GeneratedRestClient.cs (hunk 1 above)
Comment (wanlwanl) [reply src/Microsoft.Azure.SignalR.Common/Generated/GeneratedRestClient.cs]: What about this one? https://github.com/RicoSuter/NSwag/wiki/CSharpClientGenerator

I think a new script is better, since build.sh/cmd is used for generating nuget package, we don't need to wait for the whole build.

Why we need to review the auto generated code, we just need to test the api whether the behavior is expected.

File: src/Microsoft.Azure.SignalR.Common/REST/JwtTokenCredentials.cs [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    internal class JwtTokenCredentials : ServiceClientCredentials
+    {
+        private RestApiAccessTokenGenerator RestApiAccessTokenGenerator { get; }
+
+        public JwtTokenCredentials(string accessKey)
+        {
+            RestApiAccessTokenGenerator = new RestApiAccessTokenGenerator(new AccessKey(accessKey));
+        }
+
+
```
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Common/REST/JwtTokenCredentials.cs:22]: extra line

File: src/Microsoft.Azure.SignalR.Common/REST/JwtTokenCredentials.cs [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+        public override async Task ProcessHttpRequestAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            if (request == null)
+            {
+                throw new ArgumentNullException("request");
```
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Common/REST/JwtTokenCredentials.cs:27]: nameof(request)

File: src/Microsoft.Azure.SignalR.Common/REST/GeneratedRestClient.cs [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+        {
+            _productInfo = productInfo;
+        }
+
+        partial void CustomInitialize()
+        {
+            HttpClient.DefaultRequestHeaders.Add(Constants.AsrsUserAgent, _productInfo);
+        }
+
+        public static GeneratedRestClient Build(string connectionString, string productInfo)
```
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Common/REST/GeneratedRestClient.cs:24]: use `_productInfo`? remove `productInfo`?

Comment (Y-Sindo) [reply src/Microsoft.Azure.SignalR.Common/REST/GeneratedRestClient.cs:24]: This is a static method and productInfo will be passed to _productInfo?

File: src/Microsoft.Azure.SignalR.Common/REST/JwtTokenCredentials.cs (hunk 11 above)
Comment (Y-Sindo) [reply src/Microsoft.Azure.SignalR.Common/REST/JwtTokenCredentials.cs]: fix in https://github.com/Azure/azure-signalr/pull/992/commits/1ee9abc780816f67280c57052714e2e8087fa34e

File: src/Microsoft.Azure.SignalR.Common/REST/JwtTokenCredentials.cs (hunk 10 above)
Comment (Y-Sindo) [reply src/Microsoft.Azure.SignalR.Common/REST/JwtTokenCredentials.cs]: fix in https://github.com/Azure/azure-signalr/pull/992/commits/1ee9abc780816f67280c57052714e2e8087fa34e

File: src/Microsoft.Azure.SignalR.Common/REST/GeneratedRestClient.cs (hunk 12 above)
Comment (wanlwanl) [reply src/Microsoft.Azure.SignalR.Common/REST/GeneratedRestClient.cs:24]: Your goal is to `HttpClient.DefaultRequestHeaders.Add(Constants.AsrsUserAgent, _productInfo);`, is `HttpClient` static, no need to set it every time the `GeneratedRestClient` is created?

File: src/Microsoft.Azure.SignalR.Common/Generated/GeneratedRestClient.cs (hunk 1 above)
Comment (Y-Sindo) [reply src/Microsoft.Azure.SignalR.Common/Generated/GeneratedRestClient.cs]: There is already a script (client-generation.ps1)[https://github.com/Azure/azure-signalr/pull/992/files#diff-9c01492b97d08ccb869602c016adac19]

**Comment by wanlwanl** [comment]
How about exclude generated code in this PR to make it small and clean?

**Comment by Y-Sindo** [comment]
> How about exclude generated code in this PR to make it small and clean?

Generated code commits in [Add REST client generated by Autorest](https://github.com/Azure/azure-signalr/pull/992/commits/016e7e0b83b33b34373578f4c20ba3d8362e14ef), you can exclude this commit to view file changes.

File: src/Microsoft.Azure.SignalR.Common/REST/GeneratedRestClient.cs (hunk 12 above)
Comment (Y-Sindo) [reply src/Microsoft.Azure.SignalR.Common/REST/GeneratedRestClient.cs:24]: HttpClient is not a static field. It belongs to the base class of GeneratedRestClient.

**Comment by wanlwanl** [comment]
> This is a pretty large PR, how about separating it into several pieces:
> 
> 1. Only contains the generated code, and the script file to generate it, and the doc describing the steps
> 2. The convenience layer (logic layer) leveraging the generated code
> 3. Replace the current REST client with the new one

I think Liangying suggested you split this PR into 3 PRs as above. Exclude some commit isn't convenient for reviewers, especially for those who don't see your comment.

**Comment by Y-Sindo** [comment]
> > This is a pretty large PR, how about separating it into several pieces:
> > 
> > 1. Only contains the generated code, and the script file to generate it, and the doc describing the steps
> > 2. The convenience layer (logic layer) leveraging the generated code
> > 3. Replace the current REST client with the new one
> 
> I think Liangying suggested you split this PR into 3 PRs as above. Exclude some commit isn't convenient for reviewers, especially for those who don't see your comment.

Thanks for pointing out my misunderstanding. The first of the 3 PRs is #1007 .

File: src/Microsoft.Azure.SignalR.Management/ServiceManager.cs (hunk 9 above)
Comment (Y-Sindo) [reply src/Microsoft.Azure.SignalR.Management/ServiceManager.cs:162]: According to Liangying, this breaking change is acceptable.

Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Management/ServiceManager.cs:162]: does our Function output binding depend on these exceptions @zackliu? If not, is changing this exception to some general exceptions/not throw acceptable? cc @vwxyzh

Comment (wanlwanl) [reply src/Microsoft.Azure.SignalR.Management/ServiceManager.cs:162]: @vicancy It's designed for helping customer self-troubleshooting, they can catch AzureSignalRException, and the exceptions listed are having messages include some mitigate suggestions that customers can task action. Function output binding doesn't depend on them.
